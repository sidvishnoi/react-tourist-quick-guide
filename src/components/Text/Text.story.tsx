import { color, select, text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';

import * as React from 'react';
import Text, { fontSize, TextProps } from '.';

const stories = storiesOf('Typography', module);
stories.addDecorator(withKnobs);

stories.add('Typography', () => {
  const props = {
    as: select(
      'as',
      ['div', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p', 'span'],
      'p',
    ),
    color: color('color', '#000'),
    family: text('family', 'sans-serif'),
    size: select(
      'size',
      Object.keys(fontSize) as TextProps['size'][],
      'fsize_03',
    ),
  };
  return (
    <Text
      as={props.as}
      family={props.family}
      size={props.size}
      color={props.color}
    >
      {text(
        'text',
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit.' +
          ' Duis sed neque ac tortor viverra sagittis non nec massa.',
      )}
    </Text>
  );
});
