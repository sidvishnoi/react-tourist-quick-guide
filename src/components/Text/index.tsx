import * as React from 'react';
import styled from 'styled-components';

export const fontSize = {
  fsize_01: {
    size: '0.6875rem',
    lineHeight: '1rem',
  },
  fsize_02: {
    size: '0.75rem',
    lineHeight: '1.125rem',
  },
  fsize_03: {
    size: ' 0.875rem',
    lineHeight: '1.313rem',
  },
  fsize_03B: {
    size: ' 1rem',
    lineHeight: '1.5rem',
  },
  fsize_04: {
    size: '1.125rem',
    lineHeight: '1.5rem',
  },
  fsize_05: {
    size: '1.25rem',
    lineHeight: '1.625rem',
  },
  fsize_06: {
    size: '1.5rem',
    lineHeight: '2rem',
  },
  fsize_07: {
    size: '1.75rem',
    lineHeight: '2.25rem',
  },
  fsize_08: {
    size: '2.125rem',
    lineHeight: '2.75rem',
  },
  fsize_09: {
    size: '3rem',
    lineHeight: '3.5rem',
  },
};

export interface TextProps {
  as: 'div' | 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p' | 'span';
  family: string;
  color: string;
  size: keyof typeof fontSize;
  children: React.ReactNode;
  [prop: string]: any;
}

const Typography = styled.text<TextProps>`
  font-family: ${props => props.family};
  color: ${props => props.color};
  font-size: ${props => fontSize[props.size].size};
  line-height: ${props => fontSize[props.size].lineHeight};
  margin: 0.2rem 0;

  ${props =>
    (props.as === 'span' || props.as === 'div') &&
    `
    margin: 0;
  `}

  ${props =>
    props.as === 'h1' &&
    `
    padding: 1rem;
    text-align: center;
    border-top: 2px solid ${props.borderColor};
    border-bottom: 2px solid ${props.borderColor};
    margin: 0.2rem auto;
  `}
`;

Typography.defaultProps = {
  as: 'p',
  borderColor: '#000',
  color: '#000',
  family: 'serif',
  size: 'fsize_01',
};

const Text = (props: Partial<TextProps>) => {
  const { children, ...rest } = props;
  return <Typography {...rest}>{children}</Typography>;
};

export default Text;
export { Typography };
