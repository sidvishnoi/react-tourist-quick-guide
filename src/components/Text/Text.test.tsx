import { mount } from 'enzyme';
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import 'jest-styled-components';
import React = require('react');
import * as renderer from 'react-test-renderer';
Enzyme.configure({ adapter: new Adapter() });

import Text from '.';

describe('Components - Text', () => {
  it('lets use `as` to change DOMNode type', () => {
    const output = mount(<Text as="h1">PASS</Text>);
    expect(output.find('h1').exists()).toBeTruthy();
    expect(output.find('h1').text()).toEqual('PASS');
  });

  it('lets change style', () => {
    const output = mount(
      <Text as="h1" size="fsize_08">
        PASS
      </Text>,
    );
    expect(output.find('h1')).toHaveStyleRule('font-size', '2.125rem');
  });

  it('matches snapshot', () => {
    const tree = renderer.create(
      <Text as="h2" size="fsize_07" family="sans-serif">
        PASS
      </Text>,
    );
    expect(tree).toMatchSnapshot();
  });
});
