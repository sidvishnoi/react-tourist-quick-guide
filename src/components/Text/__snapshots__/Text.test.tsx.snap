// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - Text matches snapshot 1`] = `
.c0 {
  font-family: sans-serif;
  color: #000;
  font-size: 1.75rem;
  line-height: 2.25rem;
  margin: 0.2rem 0;
}

<h2
  className="c0"
  color="#000"
  size="fsize_07"
>
  PASS
</h2>
`;
